### PR TITLE
feat: Add code relates with policyMap

### PR DESCRIPTION
**Pull request by Edmond-J-A**
Add code relates with policyMap, now function hasPolicy can use variable policyMap to tell whether a policy is exist or not  directly and functions about policy can now find policy’s key directly by policyMap.
Signed-off-by: Edmond <edomondja@gmail.com>

### CODE REVIEW

File: src/model/Policy.lua [hunk 1]
```diff
@@ -346,8 +395,13 @@ function Policy:removeFilteredPolicy(sec, ptype, fieldIndex, fieldValues)
             table.insert(tmp, rule)
         end
     end
-
+    for _,policy in pairs(self.model[sec][ptype].policy) do
+        self.model[sec][ptype].policyMap[table.concat(policy,",")]=nil
```
[the remark this answers is not in the document]
Comment (Edmond-J-A) [reply src/model/Policy.lua]: Yes, it’s a better way, I have changed the code.

File: src/model/Policy.lua [hunk 2]
```diff
@@ -268,12 +301,19 @@ end
...
+        table.remove(self.model[sec][ptype].policy, i)
+        for policy,index in pairs(self.model[sec][ptype].policyMap) do
+            if index==nil then
+
+            elseif index>i then
+                self.model[sec][ptype].policyMap[policy]=self.model[sec][ptype].policyMap[policy]-1
+            elseif index==i then
+                self.model[sec][ptype].policyMap[policy]=nil
+            end
         end
```
Comment (techoner) [inline src/model/Policy.lua:315]:
```lua
local key = table.concat(rule,",")
local index = self.model[sec][ptype].policyMap[key]
table.remove(self.model[sec][ptype].policy, index)
table.remove(self.model[sec][ptype].policyMap, key)
local length = #self.model[sec][ptype].policy
for i=index, length-1, 1 do
    self.model[sec][ptype].policyMap[key]=i
end
```
Would it be more concise like this?

File: src/model/Policy.lua [hunk 3]
```diff
@@ -301,13 +341,8 @@ end
...
     for _,rule in pairs(rules) do
-        for k, v in pairs(self.model[sec][ptype].policy) do
-            if Util.arrayEquals(rule, v) then
-                table.insert(effected,rule)
-                table.remove(self.model[sec][ptype].policy, k)
-                break
-            end
-        end
+        self:removePolicy(sec,ptype,rule)
+        table.insert(effected,rule)
```
Comment (techoner) [inline src/model/Policy.lua:345]: Remove the non-existent policy, it should not be returned in `effected`.

File: src/model/Policy.lua [hunk 4]
```diff
@@ -346,8 +381,11 @@ function Policy:removeFilteredPolicy(sec, ptype, fieldIndex, fieldValues)
             table.insert(tmp, rule)
         end
     end
-
+    self.model[sec][ptype].policyMap={}
     self.model[sec][ptype].policy = tmp
+    for i,policy in pairs(self.model[sec][ptype].policy) do
+        self.model[sec][ptype].policyMap[table.concat(policy,",")]=i
+    end
```
Comment (techoner) [inline src/model/Policy.lua:388]: The for loop refers to the `removePolicy()` above .

File: src/util/Util.lua [hunk 5]
```diff
@@ -270,11 +270,19 @@ function Util.loadPolicyLine(line, model)
...
     local rules = {}
     for i = 2, #tokens do
         table.insert(rules, tokens[i])
     end
     table.insert(model.model[sec][key].policy, rules)
+    model.model[sec][key].policyMap={}
+    for i,policy in pairs(model.model[sec][key].policy) do
+        model.model[sec][key].policyMap[table.concat(policy,",")]=i
+    end
 end
```
Comment (techoner) [inline src/util/Util.lua:286]: `loadPolicyLine` is recommended to be moved to `src/persist/Adapter.lua`.

**Review by techoner — CHANGES_REQUESTED** [review]

File: src/model/Policy.lua (hunk 4 above)
Comment (qwerfdsaz) [reply src/model/Policy.lua:388]: Got it.

Comment (Edmond-J-A) [reply src/model/Policy.lua:388]: Got it.

**Pull request merged**